src/app/admin/product-form/product-form.component.ts

```diff
@@ -1,21 +1,35 @@
 import { ProductService } from './../../product.service';
 import { CategoryService } from './../../category.service';
 import { Component } from '@angular/core';
-import { Router } from '@angular/router';
+import { Router, ActivatedRoute } from '@angular/router';
+import { take } from 'rxjs/operators';
+import { Observable, of } from 'rxjs';
 
 @Component({
   selector: 'app-product-form',
   templateUrl: './product-form.component.html'
 })
 export class ProductFormComponent {
-  categoriesObservable;
+  categoriesObservable: Observable<any>;
+  userObservable: Observable<any>;
+  product = {};
 
   constructor(
     private router: Router,
+    private route: ActivatedRoute,
     private categoryService: CategoryService,
     private productService: ProductService
   ) {
     this.categoriesObservable = categoryService.getCategories();
+
+    let id = this.route.snapshot.paramMap.get('id');
+
+    if (id)
+      this.productService
+        .getProduct(id)
+        .valueChanges()
```

Comment:
the .valueChanges() should go in your service and not in this file. Better separation of concerns

Reply:
I agree.  Good catch.